dali_tf_plugin/dali_dataset_op.cc

```diff
@@ -411,42 +412,108 @@ class DALIDatasetOp::Dataset::Iterator : public DatasetIterator<Dataset> {
     return Status::OK();
   }
 
+  /**
+   * @brief Call GetNext on given input and extract the only Tensor Example that is expected as
+   * output or signall stop.
+   */
+  Status GetExampleFromInput(IteratorContext *context, int input_idx, Tensor &example,
+                             bool *end_of_sequence) {
+    TfExample input_example;
+    *end_of_sequence = false;
+    auto &input = input_impls_[input_idx];
+    // TODO(klecki): ZipDataset just goes to next iteration on Error.
+    // Desync of input datasets is not desired, we just report the problem fast
+    TF_RETURN_IF_ERROR(input->GetNext(context, &input_example, end_of_sequence));
+    if (*end_of_sequence) {
+      return Status::OK();
+    }
+
+    // Repack the single Tensor from TfExample to Batch
+    if (input_example.size() != 1) {
+      return errors::InvalidArgument("Got a sample consisting of ", input_example.size(),
+                                     " elements for input: ", input_idx,
+                                     ". Only samples of 1 element are supported.");
+    }
+    // Extract the obtained example
+    example = input_example[0];
+    return Status::OK();
+  }
+
+  /**
+   * @brief Obtain samples from input of given index, operating in batch mode
+   *
+   * out_batch will contain one output sample representing whole batch
+   */
+  Status PrepareBatchesBatchMode(IteratorContext *context, int input_idx, Batch &out_batch,
+                                 bool *end_of_sequence) {
+    int next_batch_size = dataset()->pipeline_def_.batch_size;
+    // In batch mode, we have tensors representing batches, so we keep only 1 of them
+    out_batch.clear();
+    *end_of_sequence = false;
+
+    Tensor example;
+    TF_RETURN_IF_ERROR(GetExampleFromInput(context, input_idx, example, end_of_sequence));
+    if (*end_of_sequence) {
+      return Status::OK();
+    }
+    // Batch mode, only one tensor for whole batch
+    out_batch.push_back(example);
+    return Status::OK();
+  }
+
+  /**
+   * @brief Obtain samples from input of given index, operation in sample mode.
+   *
+   * out_batch will contain sample tensors.
+   */
+  Status PrepareBatchesSampleMode(IteratorContext *context, int input_idx, Batch &out_batch,
```

Comment:
why do we have output arguments by reference and by pointer in the same API?

Reply:
Yeah, `end_of_sequence` should be a reference if it's not optional (and this function assumes that it's not null).

Reply:
That's still a weird mix of regular use of C++ and GSG.